frontend/src/hooks/useInterceptor.ts

```diff
@@ -1,31 +1,85 @@
-import { interceptor } from 'utils/request';
-import type {
-  ErrorInterceptHandler,
-  RequestInterceptHandler,
-  SuccessInterceptHandler,
-} from 'types/requestType';
-
-type UseInterceptor = {
-  accessToken: string | null;
-  onRequest: RequestInterceptHandler;
-  onSuccess: SuccessInterceptHandler;
-  onError: ErrorInterceptHandler;
+import { useCallback, useEffect } from 'react';
+import { AxiosRequestConfig } from 'axios';
+import { publicRequest, privateRequest } from 'apis/instances';
+import { TokenStatus } from 'types/userType';
+import { UserContextValues } from 'contexts/userContext';
+
+type useInterceptorProps = {
+  userState: UserContextValues;
 };
 
-const useInterceptor = ({
-  accessToken,
-  onRequest,
-  onSuccess,
-  onError,
-}: Partial<UseInterceptor>) => {
-  useEffect(() => {
-    if (accessToken !== undefined) {
-      interceptor.accessToken = accessToken;
-    }
+const useInterceptor = (customConfig: useInterceptorProps) => {
+  const { userState } = customConfig;
+
+  const onRequest = useCallback(
+    (config: AxiosRequestConfig<any>) => {
+      return {
+        ...config,
+        headers: {
+          ...config.headers,
+          Authorization: `Bearer ${userState.accessToken}`,
+        },
+      };
+    },
+    [customConfig]
+  );
+
+  const onRejected = useCallback(
+    async (error: any) => {
+      const {
+        config,
+        response: {
+          status,
+          data: { tokenStatus },
+        },
+      } = error;
+
+      if (status !== 401) {
+        return Promise.reject(error);
+      }
+
+      if (tokenStatus === TokenStatus['invalid']) {
+        userState.logout();
+        return;
+      }
+
+      if (tokenStatus === TokenStatus['expired']) {
+        const {
+          data: { accessToken },
+        } = await publicRequest('/token/refresh');
 
-    interceptor.set({ onRequest, onSuccess, onError });
-  }, [accessToken, onRequest, onSuccess, onError]);
+        const newConfig = {
+          ...config,
+          Authorization: `Bearer ${accessToken}`,
+        };
+
+        userState.setAccessToken(accessToken);
+
+        return privateRequest(newConfig);
+      }
+
+      userState.setInitialized(true);
+      return Promise.reject(error);
+    },
+    [customConfig]
+  );
+
+  const setPublicRequestInterceptor = useCallback(() => {
+    publicRequest.interceptors.response.clear();
+    publicRequest.interceptors.response.use((config) => config, onRejected);
+  }, [onRejected]);
+
+  const setPrivateRequestInterceptor = useCallback(() => {
+    privateRequest.interceptors.request.clear();
+    privateRequest.interceptors.response.clear();
+    privateRequest.interceptors.request.use(onRequest);
+    privateRequest.interceptors.response.use((config) => config, onRejected);
+  }, [onRejected, onRequest]);
+
+  useEffect(() => {
+    setPublicRequestInterceptor();
+    setPrivateRequestInterceptor();
+  }, [setPublicRequestInterceptor, setPrivateRequestInterceptor]);
```

Comment:
useCallback을 쓴 이유가 궁금하네요. 아래와 같지 않나요?

```suggestion
  useEffect(() => {
    publicRequest.interceptors.response.clear();
    publicRequest.interceptors.response.use((config) => config, onRejected);

    privateRequest.interceptors.request.clear();
    privateRequest.interceptors.response.clear();
    privateRequest.interceptors.request.use(onRequest);
    privateRequest.interceptors.response.use((config) => config, onRejected);
  }, [onRejected, onRequest]);
```

Reply:
private, public 요청 intercept를 분리하려고 했어요. 과했다는 생각이 드네요. 👍